components/ws-manager/pkg/manager/manager.go

```diff
@@ -189,27 +191,60 @@ func (m *Manager) StartWorkspace(ctx context.Context, req *api.StartWorkspaceReq
 	}
 	span.LogKV("event", "created start workspace context")
 	clog.Debug("starting new workspace")
-	// we must create the workspace pod first to make sure we don't clean up the services or configmap we're about to create
-	// because they're "dangling".
+
+	// create a Pod object for the workspace
 	pod, err := m.createWorkspacePod(startContext)
 	if err != nil {
 		return nil, xerrors.Errorf("cannot create workspace pod: %w", err)
 	}
 	span.LogKV("event", "pod description created")
-	err = m.Clientset.Create(ctx, pod)
-	if err != nil {
-		m, _ := json.Marshal(pod)
-		safePod, _ := log.RedactJSON(m)
 
-		if k8serr.IsAlreadyExists(err) {
-			clog.WithError(err).WithField("req", req).WithField("pod", safePod).Warn("was unable to start workspace which already exists")
-			return nil, status.Error(codes.AlreadyExists, "workspace instance already exists")
+	// create the Pod in the cluster and wait until is scheduled
+	// https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#workloads-that-saturate-nodes-with-pods-may-see-pods-that-fail-due-to-node-admission
+	backoff := wait.Backoff{
+		Steps:    10,
+		Duration: 100 * time.Millisecond,
+		Factor:   2.5,
+		Jitter:   0.1,
+		Cap:      5 * time.Minute,
+	}
+
+	var retryErr error
+	err = wait.ExponentialBackoff(backoff, func() (bool, error) {
+		err = m.Clientset.Create(ctx, pod)
+		if err != nil {
+			m, _ := json.Marshal(pod)
+			safePod, _ := log.RedactJSON(m)
+
+			if k8serr.IsAlreadyExists(err) {
+				clog.WithError(err).WithField("req", req).WithField("pod", safePod).Warn("was unable to start workspace which already exists")
+				return false, status.Error(codes.AlreadyExists, "workspace instance already exists")
+			}
+
+			clog.WithError(err).WithField("req", req).WithField("pod", safePod).Error("was unable to start workspace")
+			return false, err
 		}
 
-		clog.WithError(err).WithField("req", req).WithField("pod", safePod).Error("was unable to start workspace")
-		return nil, err
+		err = wait.PollWithContext(ctx, 100*time.Millisecond, 5*time.Second, podRunning(m.Clientset, pod.Namespace, pod.Name))
+		if err != nil {
+			m, _ := json.Marshal(pod)
+			safePod, _ := log.RedactJSON(m)
+			clog.WithError(err).WithField("req", req).WithField("pod", safePod).Error("was unable to reach ready state")
+			retryErr = err
+			return true, nil
```

Comment:
If we return true here, it'll stop retrying. I assume this is intentional, to avoid creating many pods which never enter a ready state?

Were you able to confirm that when an OOM kill happens with the kubelet that it occurs [within here](https://github.com/gitpod-io/gitpod/blob/45d9b3482cd12316065dbbcb11452836eddadc9e/components/ws-manager/pkg/manager/manager.go#L219-L225)?

Reply:
>If we return true here, it'll stop retrying. 

correct, since if it gets into that if clause, that means the pod is up and running now. 

>Were you able to confirm that when an OOM kill happens 

No, I spent whole day trying to repro OOM issue and was not able to. Most probably it requires high load on the cluster and some race condition for this to happen. At least I had no luck in workspace preview env, and I did try everything that I could think of.

Reply:
Hm. I think over here we actually should return `false`:
https://github.com/gitpod-io/gitpod/blob/45d9b3482cd12316065dbbcb11452836eddadc9e/components/ws-manager/pkg/manager/manager.go#L234
since if we return true, that will break exponential backoff. 
And we want to retry creating pod again if podRunning returned error.
so I think that line actually need to be changed.